roachtest/sqlsmith: internal error: ordering column group  contains non-equivalent columns (op values)

roachtest.sqlsmith/setup=rand-tables/setting=no-mutations [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/6601268?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/6601268?buildTab=artifacts#/sqlsmith/setup=rand-tables/setting=no-mutations) on release-22.2 @ [8eb9026a0ccbf3411341de62438a678fda1b64cd](https://github.com/cockroachdb/cockroach/commits/8eb9026a0ccbf3411341de62438a678fda1b64cd):


```
test artifacts and logs in: /artifacts/sqlsmith/setup=rand-tables/setting=no-mutations/run_1
	sqlsmith.go:264,sqlsmith.go:324,test_runner.go:908: error: pq: internal error: ordering column group (12,24) contains non-equivalent columns (op values)
		stmt:
		SELECT
			CASE WHEN true THEN tab_923.col1_4 ELSE tab_923.col1_3 END AS col_2150,
			e'\'\\m_yR\x0f``\b':::STRING AS col_2151,
			tab_923.col1_5 AS col_2152,
			tab_923.col1_1 AS col_2153,
			tab_923.tableoid AS col_2154,
			tab_922.col2_0 AS col_2155,
			'9037:a4c2:d96a:e9d3:4f26:b932:7b68:fceb/104':::INET AS col_2156,
			e'"\x07tC':::STRING AS col_2157,
			tab_923.col1_6 AS col_2158,
			0.9163884450683635:::FLOAT8 AS col_2159,
			tab_922.col2_0 AS col_2160,
			tab_923.col1_7 AS col_2161,
			('2010-04-21 16:01:02.000199+00:00':::TIMESTAMPTZ::TIMESTAMPTZ || e']X~e\x0e\t\x1e\x1a':::STRING::STRING)::STRING
				AS col_2162
		FROM
			defaultdb.public.table2@table2_pkey AS tab_922
			FULL JOIN defaultdb.public.table1@table1_col1_8_key AS tab_923 ON NULL
		WHERE
			false
		ORDER BY
			tab_923.col1_4 ASC, tab_922.crdb_internal_mvcc_timestamp DESC, tab_923.col1_10, tab_923.col1_7
		LIMIT
			69:::INT8;
```
<p>Parameters: <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=4</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #88648 roachtest: sqlsmith/setup=rand-tables/setting=no-mutations failed [C-test-failure O-roachtest O-robot T-sql-queries branch-master]
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=rand-tables/setting=no-mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-19888

--- a/pkg/sql/opt/ordering/ordering.go
+++ b/pkg/sql/opt/ordering/ordering.go
@@ -406,7 +406,7 @@ func checkRequired(expr memo.RelExpr, required *props.OrderingChoice) {
 	// Verify that columns in a column group are equivalent.
 	for i := range required.Columns {
 		c := &required.Columns[i]
-		if !c.Group.SubsetOf(rel.FuncDeps.ComputeEquivGroup(c.AnyID())) {
+		if !c.Group.SubsetOf(rel.FuncDeps.ComputeEquivGroup(c.AnyID())) && !rel.Cardinality.IsZero() {
 			panic(errors.AssertionFailedf(
 				"ordering column group %s contains non-equivalent columns (op %s)",
 				c.Group, expr.Op(),

--- a/pkg/sql/opt/xform/testdata/physprops/ordering
+++ b/pkg/sql/opt/xform/testdata/physprops/ordering
@@ -2974,3 +2974,61 @@ project
       │    └── projections
       │         └── true
       └── filters (true)
+
+# Regression test for #88649 - don't panic when zero-cardinality group drops
+# trivial equivalences.
+exec-ddl
+CREATE TABLE table1_88649 (
+  col1_0 FLOAT4 NOT NULL,
+  col1_1 OID NULL,
+  col1_2 STRING NOT NULL,
+  col1_3 INET NULL,
+  col1_4 INET NOT NULL,
+  col1_5 GEOGRAPHY NULL,
+  col1_6 DATE NULL,
+  col1_7 BOX2D NOT NULL,
+  col1_8 STRING NULL AS (lower(CAST(col1_1 AS STRING))) STORED,
+  col1_9 STRING NOT NULL AS (lower(CAST(col1_4 AS STRING))) STORED,
+  col1_10 FLOAT8 NOT NULL AS (col1_0 + (-2.2259812355041504):::FLOAT8) VIRTUAL,
+  col1_11 STRING NOT NULL AS (lower(CAST(col1_4 AS STRING))) VIRTUAL,
+  UNIQUE (col1_8),
+  UNIQUE (col1_4 ASC)
+)
+----
+
+exec-ddl
+CREATE TABLE table2_88649 (
+  col2_0 CHAR NOT NULL,
+  col2_1 JSONB NULL,
+  col2_2 BOX2D NOT NULL,
+  col2_3 OID NOT NULL,
+  col2_4 STRING NOT NULL AS (lower(CAST(col2_1 AS STRING))) VIRTUAL,
+  PRIMARY KEY (col2_0, col2_4 ASC, col2_2 ASC, col2_3),
+  INDEX (col2_0 ASC, col2_4) NOT VISIBLE,
+  UNIQUE (col2_3 ASC, col2_4 DESC, col2_0 ASC)
+)
+----
+
+opt
+SELECT
+  CASE WHEN true THEN tab_923.col1_4 ELSE tab_923.col1_3 END AS col_2150
+FROM
+  table2_88649@table2_88649_pkey AS tab_922
+FULL JOIN
+  table1_88649@table1_88649_col1_8_key AS tab_923 ON NULL
+WHERE
+  false
+ORDER BY
+  tab_923.col1_4 ASC, tab_922.crdb_internal_mvcc_timestamp DESC, tab_923.col1_10, tab_923.col1_7
+----
+sort
+ ├── columns: col_2150:23!null  [hidden: tab_922.crdb_internal_mvcc_timestamp:6!null col1_4:12!null]
+ ├── cardinality: [0 - 0]
+ ├── key: ()
+ ├── fd: ()-->(6,12,23)
+ ├── ordering: +(12|23),-6 [actual: ]
+ └── values
+      ├── columns: tab_922.crdb_internal_mvcc_timestamp:6!null col1_4:12!null col_2150:23!null
+      ├── cardinality: [0 - 0]
+      ├── key: ()
+      └── fd: ()-->(6,12,23)

